[DRAFT] feat: exposed composed string for psql connection

Soaib024 opened the pull request:
### Description

i) Exposed composed string for psql connection
ii) Extend complete examples to connect to database

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- When merging, use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

Soaib024 commented:
<img width="927" alt="Screenshot 2023-10-05 at 11 40 20 PM" src="https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/assets/33781996/11947fea-9878-472d-8250-dca949d3391a">


Table and rows created in postgresql instance

Soaib024 commented:
<img width="651" alt="Screenshot 2023-10-05 at 11 44 32 PM" src="https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/assets/33781996/1282b1ad-6438-48bc-af4a-d40850879600">

Exposed compose string for each service credentials

Soaib024 commented:
/run pipeline

Soaib024 commented:
/run pipeline

Soaib024 commented:
/run pipeline

Soaib024 commented:
Upgrade test is expected to [fail](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/actions/runs/6427883220/job/17454250157#step:7:2929), as this PR modifies ACL rules and also security-group for postgresql will be created using module instead of resource block in example (there isn't any changes in root module)

Soaib024 commented:
/run pipeline

Soaib024 commented:
/run pipeline

Soaib024 commented:
/run pipeline

shemau commented:
Does the test work? After running the apply locally, I see the db_connection created and complete before the VPE creation even starts. How would a connection work over private when there is no VPE gateway? (I am not sure DNS resolution works until this is live, but I may be wrong). If it does work and routes over the private network, it means that a VPE is not required/adds no value in the example. Additionally, my expectation of CBR is that they are not active immediately and the validation runs immediately after the CBR is created, possibly before it is in effect. Is there some way to confirm CBR is live before testing that the connection will run?

Soaib024 commented:
Hi @shemau, I was occupied with some other work, so I couldn't look into this for a while.

Yes, as you rightly pointed out, db_creation has to wait for the creation of `vpe`. I was testing something different when I added depends_on `module.postgresql_db` instead of `module.vpe` in a later commit. I then missed reverting it back, but I have now made the required changes.

CBR appears to be working based on my limited testing. [here](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/c5032cecf73b9b41efa1d85008d061b00695baff/examples/complete/main.tf#L165) in a different branch where Postgres is scoped to [cbr-zone](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/c5032cecf73b9b41efa1d85008d061b00695baff/examples/complete/main.tf#L203) that allows traffic from [vpc-1](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/c5032cecf73b9b41efa1d85008d061b00695baff/examples/complete/main.tf#L87) only while all other infrastructure is deployed in a different [vpc](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/c5032cecf73b9b41efa1d85008d061b00695baff/examples/complete/main.tf#L52]). As expected, this example fails after multiple retries of remote-exec where as in PR branch it executes successfully.  

Additionally, I've added `on_failure = fail` to the remote-exec, which will throw an error explicitly in case any of the commands fail to execute.

Soaib024 commented:
@daniel-butler-irl  @ocofaigh @shemau @toddgiguere 

As discussed in the deep dive session, I have separated the logic for VSI creation and added it into `tests/resources`. It still relies on `null_resources` for executing commands against the Postgres instance. However, since this is now part of `tests/resources`, I believe it should be acceptable, especially considering it's not intended for consumers. Please let me know if we should eliminate the use of `null_resources` and execute commands directly from the test itself. Additionally, I have made some changes in `example/complete` to make it work. Thank you.

[Here](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/actions/runs/6630032961/job/18012822440#step:7:3984) are the SQL query logs

Soaib024 commented:
/run pipeline

Soaib024 commented:
/run pipeline

Soaib024 commented:
<img width="1161" alt="Screenshot 2023-11-06 at 9 44 31 PM" src="https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/assets/33781996/990b7a45-6557-4e4f-ae91-1eac359c99db">

Soaib024 commented:
Upgrade test is expected to [fail](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/actions/runs/6427883220/job/17454250157#step:7:2929), as this PR modifies default  ACL rules created by vpc module and also security-group for postgresql will be created using module instead of resource block in example 

**there isn't any changes in root module**

Soaib024 commented:
/run pipeline
